x/nodes/keeper/valStateChanges.go

```diff
@@ -109,48 +109,58 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
 }
 
 // ValidateValidatorStaking - Check Validator before staking
-func (k Keeper) ValidateValidatorStaking(ctx sdk.Ctx, validator types.Validator, amount sdk.BigInt, signerAddress sdk.Address) sdk.Error {
-
-	//check the "new" validator's signature validity
-	//will recheck if validator exists
-	err, valid := ValidateValidatorMsgSigner(validator, signerAddress, k)
-	if !valid {
-		return err
+func (k Keeper) ValidateValidatorStaking(ctx sdk.Ctx,
+	validatorNew types.Validator, amount sdk.BigInt,
+	signerAddress sdk.Address) sdk.Error {
+	// check to see if the public key has already been register for that validator
+	validatorCur, found := k.GetValidator(ctx, validatorNew.Address)
+	if found &&
+		validatorNew.Address.Equals(validatorCur.Address) &&
+		validatorNew.OutputAddress != nil &&
+		k.Cdc.IsAfterNonCustodialUpgrade(ctx.BlockHeight()) &&
+		k.Cdc.IsAfterOutputAddressChangeUpgrade(ctx.BlockHeight()) {
+		// In this case, for edit stake, we skip verifying an operation
+		// against the new validator to accept a signer that appears
+		// only in the current validator.
+	} else {
+		// check the "new" validator's state first
+		err, valid := ValidateValidatorMsgSigner(validatorNew, signerAddress, k)
+		if !valid {
+			return err
+		}
 	}
 
 	//check that we don't allow nil output if we are after noncustodial upgrade
 	//so we won't accept stakes/edits with nil outputAddress
 	if k.Cdc.IsAfterNonCustodialUpgrade(ctx.BlockHeight()) {
-		if validator.OutputAddress == nil {
+		if validatorNew.OutputAddress == nil {
 			return types.ErrNilOutputAddr(k.codespace)
 		}
 	}
 
 	coin := sdk.NewCoins(sdk.NewCoin(k.StakeDenom(ctx), amount))
 
-	if int64(len(validator.Chains)) > k.MaxChains(ctx) {
+	if int64(len(validatorNew.Chains)) > k.MaxChains(ctx) {
 		return types.ErrTooManyChains(types.ModuleName)
 	}
 
-	// check to see if the public key has already been register for that validator
-	val, found := k.GetValidator(ctx, validator.Address)
 	if found {
-		//check again based on the found "state" validator
-		err, valid := ValidateValidatorMsgSigner(val, signerAddress, k)
+		//check again based on the "current" validator's state
+		err, valid := ValidateValidatorMsgSigner(validatorCur, signerAddress, k)
 		if !valid {
 			return err
 		}
 		// edit stake in 6.X upgrade
-		if ctx.IsAfterUpgradeHeight() && val.IsStaked() {
-			return k.ValidateEditStake(ctx, val, validator, amount, signerAddress)
+		if ctx.IsAfterUpgradeHeight() && validatorCur.IsStaked() {
+			return k.ValidateEditStake(ctx, validatorCur, validatorNew, amount, signerAddress)
```

Comment:
Before `ValidateEditStake`, `ValidateValidatorStaking` calls `ValidateValidatorMsgSigner` twice, one with the new validator and one with the current validator.  We need to accept a stake transaction signed by the current output address, but it doesn't pass the first `ValidateValidatorMsgSigner` against the new validator because the new validator doesn't have the signer, the current output address.  That's why we need a logic to skip the first `ValidateValidatorMsgSigner`.  We can't move that part to `ValidateEditStake`.

Well, I'd like to remove `ValidateValidatorMsgSigner` and do all checks inside `ValidateEditStake`, but that kind of refactoring change shouldn't be included in this patch.